### PR TITLE
perf: fast pq lookup

### PR DESCRIPTION
**Pull request by eddyxu**
Improve PQ lookup for about 2x by reduce the multiplications in offset calculation.

Ran `cargo bench --bench pq_dist_table`

```
5242880,L2,PQ=96,DIM=1536
                        time:   [214.81 ms 215.06 ms 215.41 ms]
                        change: [-49.469% -49.410% -49.328%] (p = 0.00 < 0.10)
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

5242880,Cosine,PQ=96,DIM=1536
                        time:   [239.96 ms 240.41 ms 241.04 ms]
                        change: [-45.693% -45.588% -45.448%] (p = 0.00 < 0.10)
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
```

Tested with

```
RUSTFLAGS="" cargo test
RUSTFLAGS="-C target-cpu=native -C target-feature=+avx512f,+avx512bw,+avx512vl" cargo +nightly test
RUSTFLAGS="-C target-cpu=native -C target-feature=+avx512f,+avx512bw,+avx512vl" cargo test. # exercise on non-avx512 branch.
```

### CODE REVIEW

File: rust/lance-index/src/vector/pq.rs [hunk 1]
```diff
@@ -267,9 +269,9 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
                             c.chunks_exact(16)
                                 .enumerate()
                                 .for_each(|(idx, lane_chunk)| {
-                                    let mut offsets: [i32; 16] = [0; 16];
+                                    let mut offsets: [i32; 16] = [(idx as i32 * 8) * 256; 16];
                                     lane_chunk.iter().enumerate().for_each(|(j, &code)| {
-                                        offsets[j] = ((idx * 8 + j) * 256 + code as usize) as i32
+                                        offsets[j] = (j * 256 + code as usize) as i32
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/pq.rs:274]: do you mean `offsets[j] +=`?

File: rust/lance-index/src/vector/pq.rs [hunk 2]
```diff
@@ -290,10 +292,11 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
...
                     c.chunks_exact(8).enumerate().for_each(|(idx, lane_chunk)| {
-                        let mut offsets: [i32; 8] = [0; 8];
-                        lane_chunk.iter().enumerate().for_each(|(j, &code)| {
-                            offsets[j] = ((idx * 8 + j) * 256 + code as usize) as i32
-                        });
+                        let mut offsets: [i32; 8] = [idx as i32 * 8 * 256; 8];
+                        lane_chunk
+                            .iter()
+                            .enumerate()
+                            .for_each(|(j, &code)| offsets[j] = (j * 256 + code as usize) as i32);
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/vector/pq.rs:299]: same here, should be `+=`?

File: rust/lance-index/src/vector/pq.rs [hunk 3]
```diff
@@ -228,10 +233,7 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
         Ok(total_distortion / data.num_rows() as f64)
     }
 
-    /// Pre-compute L2 distance from the query to all code.
-    ///
-    /// It returns the squared L2 distance.
-    fn l2_distance_table(&self, key: &dyn Array, code: &UInt8Array) -> Result<Float32Array> {
+    fn build_l2_distance_table(&self, key: &dyn Array) -> Result<Vec<f32>> {
```
Comment (westonpace) [inline rust/lance-index/src/vector/pq.rs:236]: Can you add back a comment so we know what this is creating?  Did the meaning of "distance table" change?  It used to be that distance table was distance to code points. Now it looks like it is the distance to pq centroids?

Also, it is a little odd that there is a function named `build_l2_distance_table` and `build_distance_table` and they do not mean the same thing by "distance table".

Maybe rename this method to `build_query_centroid_distances` or something else?

File: rust/lance-index/src/vector/pq.rs [hunk 4]
```diff
@@ -254,63 +253,103 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
...
+                            }
+                        } else {
+                            let mut s = 0.0;
+                            for k in 0..self.num_sub_vectors - i {
+                                *sum += distance_table
+                                    [(i + k) * num_centroids + c[vec_start + k] as usize];
+                            }
+                        }
+                    }
+                    #[cfg(not(any(feature = "nightly", target_feature = "avx512f")))]
```
Comment (westonpace) [inline rust/lance-index/src/vector/pq.rs:314]: Does this mean that `avx512f` is now the only vectorized path?  In other words, if I had avx2 but not avx512f will I see a regression?

File: rust/lance-index/build.rs [hunk 5]
```diff
@@ -7,5 +22,11 @@ fn main() -> Result<()> {
     prost_build.protoc_arg("--experimental_allow_proto3_optional");
     prost_build.compile_protos(&["./protos/index.proto"], &["./protos"])?;
 
+    let rust_toolchain = env::var("RUSTUP_TOOLCHAIN").unwrap();
+    if rust_toolchain.starts_with("nightly") {
+        // enable the 'nightly' feature flag
+        println!("cargo:rustc-cfg=feature=\"nightly\"");
```
Comment (westonpace) [inline rust/lance-index/build.rs:28]: This is cool

**Review by westonpace — APPROVED** [review]
Looks good, appreciate the new test.

Just some naming suggestions.

File: rust/lance-index/src/vector/pq.rs (hunk 4 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/pq.rs:314]: avx2 uses auto-vectorization. But for avx512f, this `gather()` is more efficient than vectorization.

File: rust/lance-index/src/vector/pq.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance-index/src/vector/pq.rs:236]: This was a sad left-over long time ago. Lemme change these function name for good this time.

**Pull request merged**